src/libmongoc/src/mongoc/mongoc-async-cmd.c

```diff
@@ -134,20 +134,36 @@ mongoc_async_cmd_run (mongoc_async_cmd_t *acmd)
 }
 
 void
-_mongoc_async_cmd_init_send (mongoc_async_cmd_t *acmd, const char *dbname)
+_mongoc_async_cmd_init_send (const mongoc_opcode_t cmd_opcode,
+                             mongoc_async_cmd_t *acmd,
+                             const char *dbname)
 {
-   acmd->ns = bson_strdup_printf ("%s.$cmd", dbname);
-
    acmd->rpc.header.msg_len = 0;
    acmd->rpc.header.request_id = ++acmd->async->request_id;
    acmd->rpc.header.response_to = 0;
-   acmd->rpc.header.opcode = MONGOC_OPCODE_QUERY;
-   acmd->rpc.query.flags = MONGOC_QUERY_SECONDARY_OK;
-   acmd->rpc.query.collection = acmd->ns;
-   acmd->rpc.query.skip = 0;
-   acmd->rpc.query.n_return = -1;
-   acmd->rpc.query.query = bson_get_data (&acmd->cmd);
-   acmd->rpc.query.fields = NULL;
+
+   acmd->ns = bson_strdup_printf ("%s.$cmd", dbname);
+
+   if (MONGOC_OPCODE_QUERY == cmd_opcode) {
+      acmd->rpc.header.opcode = MONGOC_OPCODE_QUERY;
```

Comment:
```suggestion
   if (MONGOC_OPCODE_QUERY == cmd_opcode) {
      acmd->ns = bson_strdup_printf ("%s.$cmd", dbname);
      acmd->rpc.header.opcode = MONGOC_OPCODE_QUERY;
```

`acmd->ns` is only applied to the OP_QUERY `fullCollectionName`. It should be OK to leave `acmd->ns` as NULL for OP_MSG. In `mongoc_async_cmd_destroy`, the call to `bson_free (acmd->ns);` will ignore a NULL value.